source/posix/system_info.c

```diff
@@ -236,6 +239,63 @@ void s_resolve_cmd(char *cmd, size_t len, struct aws_stack_frame_info *frame) {
 }
 #    endif
 
+int aws_backtrace(void **frames, size_t size) {
+    return backtrace(frames, size);
+}
+
+char **aws_backtrace_symbols(void *const *frames, size_t stack_depth) {
+    return backtrace_symbols(frames, stack_depth);
+}
+
+char **aws_backtrace_addr2line(void *const *stack_frames, size_t stack_depth) {
+    char **symbols = aws_backtrace_symbols(stack_frames, stack_depth);
+    AWS_FATAL_ASSERT(symbols);
+    struct aws_byte_buf lines;
+    aws_byte_buf_init(&lines, aws_default_allocator(), stack_depth * 256);
+
+    /* insert pointers for each stack entry */
+    memset(lines.buffer, 0, stack_depth * sizeof(void *));
+    lines.len += stack_depth * sizeof(void *);
+
+    /* symbols look like: <exe-or-shared-lib>(<function>+<addr>) [0x<addr>]
+     *                or: <exe-or-shared-lib> [0x<addr>]
+     * start at 1 to skip the current frame (this function) */
+    for (int frame_idx = 0; frame_idx < stack_depth; ++frame_idx) {
+        struct aws_stack_frame_info frame;
+        AWS_ZERO_STRUCT(frame);
+        const char *symbol = symbols[frame_idx];
+        if (s_parse_symbol(symbol, stack_frames[frame_idx], &frame)) {
+            goto parse_failed;
+        }
+
+        /* TODO: Emulate libunwind */
+        char cmd[sizeof(struct aws_stack_frame_info)] = {0};
+        s_resolve_cmd(cmd, sizeof(cmd), &frame);
+        FILE *out = popen(cmd, "r");
+        if (!out) {
+            goto parse_failed;
+        }
+        char output[1024];
+        if (fgets(output, sizeof(output), out)) {
+            /* if addr2line or atos don't know what to do with an address, they just echo it */
+            /* if there are spaces in the output, then they resolved something */
+            if (strstr(output, " ")) {
+                symbol = output;
+            }
+        }
+        pclose(out);
+
+    parse_failed:
+        /* record the pointer to where the symbol will be */
+        *((char **)&lines.buffer[frame_idx * sizeof(void *)]) = (char *)lines.buffer + lines.len;
+        struct aws_byte_cursor line_cursor = aws_byte_cursor_from_c_str(symbol);
+        line_cursor.len += 1; /* strings must be null terminated, make sure we copy the null */
+        aws_byte_buf_append_dynamic(&lines, &line_cursor);
+    }
+    free(symbols);
+    return (char **)lines.buffer; /* caller is responsible for freeing */
```

Comment:
This makes me really nervous, it goes against all of the safety mechanisms we have in place. Can we not just return the byte_buf instead? Then the user just calls `aws_byte_buf_clean_up`

Reply:
I considered that. The reason I didn't is that it's an array of strings. The user would have to unpack it from the byte_buf manually, by grabbing ```buf.buffer``` and casting it to a ```const char**```. Also, the calls we are wrapping/emulating/replacing work this way. Thoughts?

Reply:
I suppose that makes sense.

The only thing I can think to do that would make the byte_buf version work better would be to add a `aws_byte_buf_write_ptr()` and `aws_byte_cursor_read_ptr()`, but that seems like a bad pattern to enable. Let's leave as is.